pkg/tnf/handlers/logging/logging_test.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/test-network-function/test-network-function/pkg/tnf/reel"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+const (
+	testTimeoutDuration = time.Second * 5
+)
+
+var (
+	genericTestSchemaFile = path.Join("schemas", "generic-test.schema.json")
+	loggingFilename       = "logging.json"
+	/* #nosec G101 */
+	expectedPassPattern  = "(?m)[1-9]"
+	expectedFailPattern  = "(?m)[0]"
+	expectedErrPattern   = "(?m).*"
+	pathRelativeToRoot   = path.Join("..", "..", "..", "..")
+	pathToTestSchemaFile = path.Join(pathRelativeToRoot, genericTestSchemaFile)
+	testPodNameSpace     = "testnamespace"
+	testPodName          = "testPodname"
+)
+
+func createTest() (*tnf.Tester, []reel.Handler, *gojsonschema.Result, error) {
+	values := make(map[string]interface{})
+	values["POD_NAMESPACE"] = testPodNameSpace
+	values["POD_NAME"] = testPodName
+	return generic.NewGenericFromMap(loggingFilename, pathToTestSchemaFile, values)
+}
+
+func TestLogging_Args(t *testing.T) {
+	test, handlers, jsonParseResult, err := createTest()
+
+	assert.Nil(t, err)
+	assert.True(t, jsonParseResult.Valid())
+	assert.NotNil(t, handlers)
+
+	assert.Nil(t, (*test).Args())
+}
+
+func TestLogging_GetIdentifier(t *testing.T) {
+	test, handlers, jsonParseResult, err := createTest()
+
+	assert.Nil(t, err)
+	assert.True(t, jsonParseResult.Valid())
+	assert.NotNil(t, handlers)
+
+	assert.Equal(t, identifier.LoggingURLIdentifier, (*test).GetIdentifier())
+}
+
+func TestLogging_ReelFirst(t *testing.T) {
+	_, handlers, jsonParseResult, err := createTest()
+
+	assert.Nil(t, err)
+	assert.True(t, jsonParseResult.Valid())
+	assert.NotNil(t, handlers)
+
+	assert.Equal(t, 1, len(handlers))
+	handler := handlers[0]
+	step := handler.ReelFirst()
+	expectedCommand := fmt.Sprintf("oc logs -n %s %s --tail 5 | wc -l", testPodNameSpace, testPodName)
+	assert.Equal(t, expectedCommand, step.Execute)
+	assert.Contains(t, step.Expect, expectedPassPattern, expectedFailPattern, expectedErrPattern)
+	assert.Equal(t, testTimeoutDuration, step.Timeout)
+}
+
+func TestLogging_ReelEof(t *testing.T) {
+	_, handlers, jsonParseResult, err := createTest()
+
+	assert.Nil(t, err)
+	assert.True(t, jsonParseResult.Valid())
+	assert.NotNil(t, handlers)
+
+	assert.Equal(t, 1, len(handlers))
+	handler := handlers[0]
+	// just ensure there isn't a panic
+	handler.ReelEOF()
+}
+
+func TestLogging_ReelTimeout(t *testing.T) {
+	_, handlers, jsonParseResult, err := createTest()
+
+	assert.Nil(t, err)
+	assert.True(t, jsonParseResult.Valid())
+	assert.NotNil(t, handlers)
+
+	assert.Equal(t, 1, len(handlers))
+	handler := handlers[0]
+	assert.Nil(t, handler.ReelTimeout())
+}
+
+func TestLogging_ReelMatch(t *testing.T) {
+	tester, handlers, jsonParseResult, err := createTest()
+
+	assert.Nil(t, err)
+	assert.True(t, jsonParseResult.Valid())
+	assert.NotNil(t, handlers)
+
+	assert.Equal(t, 1, len(handlers))
+	handler := handlers[0]
+	step := handler.ReelMatch(expectedPassPattern, "", "4")
```

Comment:
Can we unit test the whole matching process ? with ```ReelMatch``` we can only match one pattern at a time.

Reply:
Another generic question is :
there's some discrepancy between json schema definition and the code
in tnf/test.go
```
        ERROR = iota
	// SUCCESS represents a successful test.
	SUCCESS
	// FAILURE represents a failed test.
	FAILURE
ERROR = 0; SUCCESS = 1; FAILURE = 2
```
Yet in JSON generic schema
```
0 indicates SUCCESS, 1 indicates FAILURE, 2 indicates ERROR."
```
is this a bug ? or there's a layer that translate the JSON to the right value ?

in all cases, I think we should align the error code between json and the code itself.

Reply:
@ryandgoulding Could you please take a look at Salah's questions?

Reply:
> 0 indicates SUCCESS, 1 indicates FAILURE, 2 indicates ERROR.

The comment is stale.  0 indicates ERROR, 1 indicates SUCCESS, 2 indicates FAILURE.

Reply:
> Can we unit test the whole matching process ? with `ReelMatch` we can only match one pattern at a time.
> For example, in this case, I'm able to test only every pattern alone and can't test the "if else" logic embedded in json definition.

I do not understand what you are asking?  You can test `ReelMatch` from a Unit test context many times, creating a new instance of the Handler/Test each time to reset any localized state.  In each consecutive call, pass the arguments you wish to test.

Reply:
my comment is about testing any input and see if it gives the right result at the end. and not testing single pattern against single entry in ReelMatch()

so for example, in my logging.json I have this defined 
```
resultContexts":[
    {
      "pattern": "(?m)[1-9]",
      "defaultResult": 1
    },
    {
      "pattern": "(?m)[0]",
      "defaultResult": 2
    },
    {
      "pattern": "(?m)[.*]",
      "defaultResult": 0
    }
  ]
  ```
  and in the UT I have some tests like -in here, we're testing a single input with a single pattern-
  ```
  handler.ReelMatch(expectedFailPattern, "", "0")
  handler.ReelMatch(expectedErrPattern, "", "a")
  ```
  But I don't have test for the whole logic. By Whole logic, I mean when I send an input, it will return an expected result.
  For example, nothing prevents me from doing something like this :
  ```
resultContexts":[
    {
      "pattern": "(?m)[1-9]",
      "defaultResult": 1
    },
    {
      "pattern": "(?m)[1-9]",
      "defaultResult": 2
    }
  ]
  ```
  if I run both UT, they will pass
   ```
  handler.ReelMatch(expectedPassPattern, "", "0")
  handler.ReelMatch(expectedErrPattern, "", "0")
  ```

what I want, is something like this :

```
input :="a"
assert(my_function("a")==tnf.ERROR)
```
 
not sure if my comment is still unclear. This is my first PR in this project. so there maybe something I'm saying that does not make much sense

Reply:
```
resultContexts":[
  {
    "pattern": "(?m)[1-9]",
    "defaultResult": 1
  },
  {
    "pattern": "(?m)[1-9]",
    "defaultResult": 2
  }
]
```

You should not repeat a pattern regex;  each regex must be distinct to properly capture the output.  Additionally, element order is determined by the ordering in the `*reel.Step.Expect` slice returned  by `ReelFirst()`.  

```
where my_function includes all the logic that parse the input and gives a final result
```

This isn't possible;  `my_function` is bound to the `reel.Handler` at hand (i.e., it is an instance method).  It is not a static function.  You should create a new Handler per test-input.

Reply:
Thanks for the explanation.